vertical-pod-autoscaler/pkg/admission-controller/logic/server_test.go

```diff
@@ -286,3 +255,131 @@ func TestGetPatchesForResourceRequest(t *testing.T) {
+			vpa:         vpa_types.VerticalPodAutoscaler{},
+			isCreate:    true,
+			expectError: fmt.Errorf("TargetRef is required. If you're using v1beta1 version of the API, please migrate to v1beta2."),
+		},
+		{
+			name: "no update mode",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					UpdatePolicy: &vpa_types.PodUpdatePolicy{},
+				},
+			},
+			expectError: fmt.Errorf("UpdateMode is required if UpdatePolicy is used"),
+		},
+		{
+			name: "bad update mode",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					UpdatePolicy: &vpa_types.PodUpdatePolicy{
+						UpdateMode: &badUpdateMode,
+					},
+				},
+			},
+			expectError: fmt.Errorf("unexpected UpdateMode value bad"),
+		},
+		{
+			name: "no policy name",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					ResourcePolicy: &vpa_types.PodResourcePolicy{
+						ContainerPolicies: []vpa_types.ContainerResourcePolicy{{}},
+					},
+				},
+			},
+			expectError: fmt.Errorf("ContainerPolicies.ContainerName is required"),
+		},
+		{
+			name: "invalid scaling mode",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					ResourcePolicy: &vpa_types.PodResourcePolicy{
+						ContainerPolicies: []vpa_types.ContainerResourcePolicy{
+							{
+								ContainerName: "loot box",
+								Mode:          &badPolicyMode,
+							},
+						},
+					},
+				},
+			},
+			expectError: fmt.Errorf("unexpected Mode value bad"),
+		},
+		{
+			name: "bad limits",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					ResourcePolicy: &vpa_types.PodResourcePolicy{
+						ContainerPolicies: []vpa_types.ContainerResourcePolicy{
+							{
+								ContainerName: "loot box",
+								MinAllowed: apiv1.ResourceList{
+									cpu: resource.MustParse("100"),
+								},
+								MaxAllowed: apiv1.ResourceList{
+									cpu: resource.MustParse("10"),
+								},
+							},
+						},
+					},
+				},
+			},
+			expectError: fmt.Errorf("max resource for cpu is lower than min"),
+		},
+		{
+			name: "all valid",
+			vpa: vpa_types.VerticalPodAutoscaler{
+				Spec: vpa_types.VerticalPodAutoscalerSpec{
+					ResourcePolicy: &vpa_types.PodResourcePolicy{
+						ContainerPolicies: []vpa_types.ContainerResourcePolicy{
+							{
+								ContainerName: "loot box",
+								Mode:          &validScalingMode,
+								MinAllowed: apiv1.ResourceList{
+									cpu: resource.MustParse("10"),
+								},
+								MaxAllowed: apiv1.ResourceList{
+									cpu: resource.MustParse("100"),
+								},
+							},
+						},
+					},
+					UpdatePolicy: &vpa_types.PodUpdatePolicy{
+						UpdateMode: &validUpdateMode,
+					},
+				},
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("test case: %s", tc.name), func(t *testing.T) {
+			err := validateVPA(&tc.vpa, tc.isCreate)
+			if tc.expectError == nil {
```

Comment:
I wonder, isn't it simply assert.Equal(t, err, tc.expectError)?

Reply:
No, unfortunately fmt.Errorf("") == fmt.Errorf("") is false.

Reply:
I just tried assert.Equal(t, fmt.Errorf(""), fmt.Errorf("")) and it works